src/pkg/cli/compose_validation.go

```diff
@@ -169,86 +169,89 @@ func validateProject(project *compose.Project) error {
-				return fmt.Errorf("unsupported compose directive: deploy mode: %q", svccfg.Deploy.Mode)
-			}
-			if len(svccfg.Deploy.Labels) > 0 {
-				warnf("unsupported compose directive: deploy labels")
+				return fmt.Errorf("service %q: unsupported compose directive: deploy mode: %q", svccfg.Name, svccfg.Deploy.Mode)
 			}
 			if svccfg.Deploy.UpdateConfig != nil {
-				return fmt.Errorf("unsupported compose directive: deploy update_config")
+				return fmt.Errorf("service %q: unsupported compose directive: deploy update_config", svccfg.Name)
 			}
 			if svccfg.Deploy.RollbackConfig != nil {
-				return fmt.Errorf("unsupported compose directive: deploy rollback_config")
+				return fmt.Errorf("service %q: unsupported compose directive: deploy rollback_config", svccfg.Name)
 			}
 			if svccfg.Deploy.RestartPolicy != nil {
-				return fmt.Errorf("unsupported compose directive: deploy restart_policy")
-			}
-			if len(svccfg.Deploy.Placement.Constraints) != 0 || len(svccfg.Deploy.Placement.Preferences) != 0 || svccfg.Deploy.Placement.MaxReplicas != 0 {
-				warnf("unsupported compose directive: deploy placement")
+				return fmt.Errorf("service %q: unsupported compose directive: deploy restart_policy", svccfg.Name)
 			}
 			if svccfg.Deploy.EndpointMode != "" {
-				return fmt.Errorf("unsupported compose directive: deploy endpoint_mode")
+				return fmt.Errorf("service %q: unsupported compose directive: deploy endpoint_mode", svccfg.Name)
 			}
 			if svccfg.Deploy.Resources.Limits != nil && svccfg.Deploy.Resources.Reservations == nil {
-				warnf("no reservations specified; using limits as reservations")
+				warnf("service %q: no reservations specified; using limits as reservations", svccfg.Name)
 			}
 			reservations := getResourceReservations(svccfg.Deploy.Resources)
-			if reservations.NanoCPUs < 0 { // "0" just means "as small as possible"
-				return fmt.Errorf("invalid value for cpus: %v", reservations.NanoCPUs)
+			if reservations != nil && reservations.NanoCPUs < 0 { // "0" just means "as small as possible"
+				return fmt.Errorf("service %q: invalid value for cpus: %v", svccfg.Name, reservations.NanoCPUs)
+			}
+			if reservations == nil || reservations.MemoryBytes == 0 {
+				warnf("service %q: missing memory reservation; specify deploy.resources.reservations.memory to avoid out-of-memory errors", svccfg.Name)
+			}
+			if len(svccfg.Deploy.Labels) > 0 {
+				warnf("service %q: unsupported compose directive: deploy labels", svccfg.Name)
+			}
+			if len(svccfg.Deploy.Placement.Constraints) != 0 || len(svccfg.Deploy.Placement.Preferences) != 0 || svccfg.Deploy.Placement.MaxReplicas != 0 {
+				warnf("service %q: unsupported compose directive: deploy placement", svccfg.Name)
 			}
-		}
-		var reservations *compose.Resource
-		if svccfg.Deploy != nil {
-			reservations = getResourceReservations(svccfg.Deploy.Resources)
-		}
-
-		if svccfg.Deploy == nil || reservations == nil || reservations.MemoryBytes == 0 {
```

Comment:
It seems we are removing the memory reservation warning if `deploy==nil`, i guess this never happens too?